### PR TITLE
Want some more user-callbacks?

### LINKED ISSUE #712: Some callbacks are misgenerated
See https://github.com/gtk-rs/gtk/pull/770#issuecomment-458495560 and comments on the specific code with explanations right below.

@GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #712 

cc @sdroege @EPashkin

### CODE REVIEW

File: src/codegen/function_body_chunk.rs [hunk 1]
```diff
@@ -474,7 +474,13 @@ impl Builder {
                         match rust_type(env, trampoline.ret.typ).unwrap().as_str() {
                             "GString" |
                             "File" => "res.to_glib_full()".to_owned(),
-                            _ => "res.to_glib()".to_owned(),
+                            _ => {
+                                if env.type_(trampoline.ret.typ).is_fundamental_type(env) {
```
Comment (sdroege) [inline src/codegen/function_body_chunk.rs]: Are enums, flags and other scalar types also considered "fundamental"?

**Comment by sdroege** [comment]
Does this cause any changes in the gtk-rs crates?

**Comment by GuillaumeGomez** [comment]
It makes the code work but that's pretty much all. No changes on the `to_glib`/`to_glib_full`.

**Comment by sdroege** [comment]
Seems good to me then apart from the comment above :)

File: src/codegen/function_body_chunk.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/codegen/function_body_chunk.rs]: No.

Comment (sdroege) [reply src/codegen/function_body_chunk.rs]: Then this is wrong and you need to match more on the actual type (what is returned by `env.type_()`).

There should be a function already for figuring this out.

Comment (sdroege) [reply src/codegen/function_body_chunk.rs]: `impl TrampolineToGlib for library::Parameter` maybe, or code related to that

Comment (GuillaumeGomez) [reply src/codegen/function_body_chunk.rs]: Ok!

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by sdroege** [comment]
Heh, so much simpler. Looks good to me then as long as it doesn't change any other code in the gtk-rs crates apart from the broken ones :)

**Comment by EPashkin** [comment]
Confirmed that it only fixed gtk in gtk-rs crates.
@GuillaumeGomez, @sdroege Many thanks to both of you.

@sdroege strange, but listbox_model is broken again.
```
error[E0282]: type annotations needed
   --> src\bin\listbox_model.rs:30:28
    |
30  |       (@param $x:ident) => ( $x );
    |                              ^^ consider giving this closure parameter a type
...
83  |       listbox.bind_model(&model, clone!(window_weak => move |item| {
    |  ________________________________-
84  | |         let box_ = gtk::ListBoxRow::new();
85  | |         let item = item.downcast_ref::<RowData>().expect("Row data is of wrong type");
86  | |
...   |
162 | |         box_
163 | |     }));
    | |______- in this macro invocation
    |
    = note: type must be known at this point
```

**Comment by sdroege** [comment]
> @sdroege strange, but listbox_model is broken again.

Yes, that's https://github.com/gtk-rs/gir/issues/713

**Pull request merged**